### PR TITLE
adding Vfied

### PR DESCRIPTION
**Pull request by h-haddad**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 1]
```diff
@@ -15,6 +15,2538 @@
...
+              getForm().startProtected(null)
+            }
+            val query = Query(getForm().dBContext.defaultConnection)
+
+            query.addString(list!!.getColumn(0).column)
+            query.addString(evalListTable())
+            query.addString(getSql(block!!.getActiveRecord()))
+            query.open(SELECT_IS_IN_LIST)
+            exists = query.next()
+            query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1548]: you won't need to introduce a new variable query if you use let  : 
Query(getForm().dBContext.defaultConnection).let {
              it.addString(list!!.getColumn(0).column)
              it.addString(evalListTable())
              it.addString(getSql(block!!.getActiveRecord()))
              it.open(SELECT_IS_IN_LIST)
              exists = it.next()
              it.close()
            }

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 2]
```diff
@@ -15,6 +15,2538 @@
...
+        }
+      } catch (e: Throwable) {
+        throw VExecFailedException(e)
+      }
+      if (!exists) {
+        throw VFieldException(this, MessageCode.getMessage("VIS-00001"))
+      }
+      return
+    } else {
+      var query: Query
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1581]: Delete

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 3]
```diff
@@ -15,6 +15,2538 @@
...
+            if (!query.next()) {
+              count = 0
+            } else {
+              count = 1
+              result = query.getString(1)
+              if (query.next()) {
+                count = 2
+              }
+            }
+            query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1609]: use  Query(getForm().dBContext.defaultConnection).let {.. }

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 4]
```diff
@@ -15,6 +15,2538 @@
...
+          getForm().startProtected(null)
+          val query = Query(getForm().dBContext.defaultConnection)
+          query.addString(list!!.getColumn(0).column)
+          query.addString(evalListTable())
+          query.addString(getSql(block!!.getActiveRecord()))
+          query.open(SELECT_IS_IN_LIST)
+          if (query.next()) {
+            id = query.getInt(1)
+          }
+          query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1700]: use let{}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 5]
```diff
@@ -15,6 +15,2538 @@
...
+                      " WHERE    $3 = " + selected
+              val query = Query(getForm().dBContext.defaultConnection)
+
+              query.addString(list!!.getColumn(0).column)
+              query.addString(evalListTable())
+              query.addString("ID")
+              query.open(SELECT_IS_IN_LIST)
+              query.next()
+              result = query.getObject(1)
+              query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1808]: Use let {}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 6]
```diff
@@ -15,6 +15,2538 @@
...
+          query.addString(evalListTable())
+          query.addString("ID")
+          query.addInt(id)
+          query.open("SELECT $1 FROM $2 WHERE $3 = #4")
+          result = if (query.next()) {
+            query.getObject(1)
+          } else {
+            null
+          }
+          query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:2074]: use let{}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt (hunk 3 above)
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1609]: use let{}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 7]
```diff
@@ -15,6 +15,2538 @@
...
+
+    val newForm: VDictionary? = if (list!!.getNewForm() != null) {
+      // OLD SYNTAX
+      Module.getExecutable(list!!.getNewForm()) as VDictionary
+    } else if (list!!.getAction() != -1) {
+      // NEW SYNTAX
+      block!!.executeObjectTrigger(list!!.getAction()) as VDictionary?
+    } else {
+      null // should never happen.
+    }
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1735]: Use when{}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 8]
```diff
@@ -15,6 +15,2538 @@
...
+              continue
+            }
+            var i = 0
+            while (i < lines.size) {
+              lines[i][lineCount] = query.getObject(i + if (SKIP_FIRST_COLUMN) 2 else 1)
+              i += 1
+            }
+            lineCount += 1
+          }
+          query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1755]: use let

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 9]
```diff
@@ -15,6 +15,2538 @@
...
+    while (true) {
+      try {
+        getForm().startProtected(null)
+        val query = Query(getForm().dBContext.defaultConnection)
+
+        query.open(qrybuf)
+        while (value == null && query.next()) {
+          value = query.getObject(1)
+        }
+        query.close()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1896]: use let

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 10]
```diff
@@ -15,6 +15,2538 @@
...
+      }
+    }
+  }
+
+  /**
+   * Checks that field value exists in list
+   */
+  protected fun selectFromList(gotoNextField: Boolean) {
+    val qrybuf = StringBuffer()
+
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1834]: you can use : 
val qrybuf = buildString{...} and remove qrybuf occurencies inside the scope

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 11]
```diff
@@ -15,6 +15,2538 @@
...
+   * and the object values of the suggestions.
+   * @throws VException Visual exceptions related to database errors.
+   */
+  fun getSuggestions(query: String?): Array<Array<String>>? {
+    return if (query == null || autocompleteType == VList.AUTOCOMPLETE_NONE) {
+      null
+    } else {
+      val suggestions: MutableList<Array<String>>
+      val qrybuf: StringBuffer = StringBuffer()
+
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1939]: use buildString{}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 12]
```diff
@@ -15,6 +15,2538 @@
...
+          // default should never reached
+          qrybuf.append(" = ")
+          qrybuf.append(Utils.toSql(query))
+        }
+      }
+      qrybuf.append(" ORDER BY 1")
+      while (true) {
+        try {
+          getForm().startProtected(null)
+          val sqlQuery = Query(getForm().dBContext.defaultConnection)
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1973]: use let{}

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 13]
```diff
@@ -15,6 +15,2538 @@
...
+   * prepare a snapshot
+   *
+   * @param     fieldPos        position of this field within block visible fields
+   */
+  fun prepareSnapshot(fieldPos: Int, active: Boolean) {
+    // !!! TO DO
+  }
+
+  override fun toString(): String {
+    val information = StringBuffer()
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:2173]: buildString{}

**Review by ZMarouani — CHANGES_REQUESTED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt (hunk 2 above)
Comment (h-haddad) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1581]: cannot delete it

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt (hunk 3 above)
Comment (h-haddad) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1609]: query used in if (!query.next()) {

Comment (h-haddad) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:1609]: same

**Review by ZMarouani — APPROVED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 14]
```diff
@@ -18,162 +18,2053 @@
...
+   */
+  fun getSearchCondition(): String? {
+    return if (isNull(block!!.activeRecord)) {
+      when(getSearchOperator()) {
+        VConstants.SOP_EQ -> null
+        VConstants.SOP_NE -> "IS NOT NULL"
+        else -> "IS NULL"
+      }
+    } else {
+      var operator: String = VConstants.OPERATOR_NAMES[getSearchOperator()]
```
Comment (SNsir) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:670]: var operator = VConstants.OPERATOR_NAMES[getSearchOperator()]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 15]
```diff
@@ -18,162 +18,2053 @@
...
+
+    val defaultAccess = getDefaultAccess()
+    var accessTemp = defaultAccess //dynAccess;
+
+    if (current != -1) {
+      if (!block!!.isRecordInsertAllowed(current)) {
+        accessTemp = VConstants.ACS_SKIPPED
+      } else if (hasTrigger(VConstants.TRG_FLDACCESS)) {
+        // evaluate ACCESS-Trigger
+        val oldRow: Int = block!!.activeRecord
```
Comment (SNsir) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:550]: val oldRow = block!!.activeRecord

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 16]
```diff
@@ -18,162 +18,2053 @@
...
+        val old: VField = block!!.getActiveField()!!
```
Comment (SNsir) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:551]: val old = block!!.getActiveField()!!

**Review by SNsir — CHANGES_REQUESTED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 17]
```diff
@@ -18,162 +18,2053 @@
...
+  /**
+   * when leaving field, if text is okay, set value in record
+   * @exception VException      an exception is raised if text is bad
+   */
+  fun leave(check: Boolean) {
+    assert(this === block!!.getActiveField()) { threadInfo() + "current field: " + block!!.getActiveField() }
+    try {
+      if (check && changed) {
+        if (changedUI && hasListener) {
+          checkType(getDisplayedValue(true)!!)
```
[the remark this answers is not in the document]
Comment (SNsir) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:449]: change it in line 467 also

**Review by SNsir — APPROVED** [review]

**Pull request merged**